File: _95_UniqueBinarySearchTreesII.py
from typing import List
from typing import Optional
from pytest import mark

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def generateTrees(n: int) -> List[Optional[TreeNode]]:
        start, end = 1, n
        
        def RecurTree(start: int, end: int):
            if start > end:
                return [None]
            if start == end:
                return [TreeNode(start)]
            result = []
            for i in range(start, end + 1):
                left = RecurTree(start, i - 1)
                right = RecurTree(i + 1, end)
                for l in left:
                    for r in right:
                        tmp = TreeNode(i)
                        tmp.left = l
                        tmp.right = r
                        result.append(tmp)
            return result

        return RecurTree(start, end)

File: _95_UniqueBinarySearchTreesII.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private LinkedList<TreeNode> generateLinkedTrees(int left, int right) {
        LinkedList<TreeNode> allTrees = new LinkedList<>();
        if (left > right) {
            // In fact this func will return allTrees here, since it won't
            // enter the next for loop because left > right.
            // The null is for corner case when left or right is chosen
            // as root, and they will lack either left or right children
            allTrees.add(null);
        }
        for (int i = left; i <= right; i++) {
            LinkedList<TreeNode> leftTrees = generateLinkedTrees(left, i - 1);
            LinkedList<TreeNode> rightTrees = generateLinkedTrees(i + 1, right);
            for (TreeNode l : leftTrees) {
                for (TreeNode r : rightTrees) {
                    TreeNode root = new TreeNode(i);
                    root.left = l;
                    root.right = r;
                    allTrees.add(root);
                }
            }
        }
        return allTrees;
    }
    public List<TreeNode> generateTrees(int n) {
        return generateLinkedTrees(1, n);
    }
}